### PR TITLE
[28017] Integration tests for win_rabbitmq_plugin

### PR DESCRIPTION
**Pull request by ar7z1**
##### SUMMARY

This PR complements #28017: I added integration tests for `win_rabbitmq_plugin` module.

Additionally I added some code to search RabbitMQ installation path from Windows Registry. This allows to skip adding `prefix` option or adding `C:\Program Files\RabbitMQ Server\rabbitmq_server-<version>\sbin` into your `PATH`. I used `Get-RabbitMQPath` function from [rabbitmq chocolatey package](https://github.com/rabbitmq/chocolatey-package/blob/master/tools/chocolateyHelpers.ps1#L22) as an example.

Tests requires a [rabbitmq chocolatey package](https://chocolatey.org/packages/rabbitmq) to be installed:
```yaml
win_chocolatey:
   name: rabbitmq
   state: present
```
This package sometimes fails to install on 32-bit Windows. I created a [PR](https://github.com/rabbitmq/chocolatey-package/pull/3) to rabbitmq chocolatey repo, but it is still not merged yet.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

win_rabbitmq_plugin

##### ANSIBLE VERSION

```
ansible 2.4.0 (28017-tests-for-win_rabbitmq_plugin cac520329c) last updated 2017/08/13 13:29:54 (GMT +300)
  config file = /home/art/.ansible.cfg
  configured module search path = [u'/home/art/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/art/projects/github/ansible/lib/ansible
  executable location = /home/art/projects/github/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by ar7z1** [comment]
@dagwieers Can you take a look at this PR?

**Comment by ar7z1** [comment]
It seems there is some problem to install RabbitMQ on Windows Server 2008 SP2. :-(

Two builds ([this](https://app.shippable.com/github/ansible/ansible/runs/33100/10/console) and [this](https://app.shippable.com/github/ansible/ansible/runs/33101/10/console)) failed during RabbitMQ installation.

**Comment by ar7z1** [comment]
As you can see, integration tests failed by timeout on Windows 2008 SP2. And they failed on task:
```yaml
- name: Ensure RabbitMQ installed
  win_chocolatey:
    name: rabbitmq
    state: present
```

I reproduced the issue on my Windows Server 2008 SP2 and investigated the problem.

RabbitMQ chocolatey installer [installs erlang](https://chocolatey.org/packages/rabbitmq) as a dependency. Erlang installer tries to install ["Visual C++ Redistributable Packages for Visual Studio 2013"](https://www.microsoft.com/en-us/download/details.aspx?id=40784). And vcredist installer never quits. :-(

All vcredist installer threads hangs in state "Wait:WrUserRequest":
wait

And this is occurred only when tries to install vcredist over WinRM!


After a lot of hours of debugging, I found the issue in Wix bootstrapper (it's used in vcredist.exe):
* [Discussion 1](https://www.mail-archive.com/wix-devs@lists.sourceforge.net/msg06150.html) and [discussion 2](http://windows-installer-xml-wix-toolset.687559.n2.nabble.com/wixstdba-based-installer-hangs-when-invoked-by-winrm-td7596955.html).
* [Bug to Microsoft](https://connect.microsoft.com/VisualStudio/feedback/details/783880), that was ignored (just closed).
* [PR with fix](https://github.com/wixtoolset/wix3/pull/157) to Wix github repo.

The main problem is that I can't fix vcredist installer (I need to update Wix bootstrapper in it). :-(

Possible solutions are:
* Create a PR to Erlang to use a newer vcredist installer. That's quite hard. :-(
* Make an ugly hack in tests which will be:
  1. Start rabbitmq installation via chocolatey.
  2. Periodically get the list of running processes and if detect hanging vcredist installation process, kill it. Someting like this:
    ```powershell
    function Get-Pids($processName)
    {
        $pids = ps | where {$_.Name -eq $processName} | select -ExpandProperty Id
        return ,$pids
    }

    function Are-AllThreadsWait($processId)
    {
        $process=[System.Diagnostics.Process]::GetProcessById($processId)
        return ($process.Threads | where { $_.ThreadState -ne 'Wait' -or $_.WaitReason -ne 'UserRequest' }).Count -eq 0
    }

    $ErrorActionPreference = "Stop"

    foreach ($p in (Get-Pids -processName 'vcredist_x86')) {
        if (Are-AllThreadsWait -processId $p) {
            Stop-Process -Id $p
        }
    }
    ```
    And it's too not so easy to do it via ansible:
    * [win_scheduled_task](http://docs.ansible.com/ansible/latest/win_scheduled_task_module.html) does not support Windows 2008.
   * There are some problems to start `async` `win_chocolatey` task and poll it: I need to call `async_status` and `win_shell` simultaneously.

**Comment by dagwieers** [comment]
@ar7z1 So, yes, skip 2008R2 for now from the integration tests, open an issue to the Erlang project for fixing/upgrading their installer (but we cannot wait for it, it is just a courtesy to them) and leave it at that.

Maybe add a TODO: message in the integration tests to revise this at some later point and add a reference to the open issue at Erlang (and this discussion as well).

File: test/integration/targets/win_rabbitmq_plugin/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Setup action creates ansible_distribution_version variable
+- action: setup
+
+- include_tasks: tasks/tests.yml
+  # Works on windows >= Windows 7/Windows Server 2008 R2
+  when: ansible_distribution_version | version_compare('6.1', '>=')
```
Comment (ar7z1) [inline test/integration/targets/win_rabbitmq_plugin/tasks/main.yml]: [Windows versions](https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms724832%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396).

**Comment by ar7z1** [comment]
@dagwieers I [skipped](https://github.com/ansible/ansible/pull/28118/commits/7024b34ba5d12b9ed341d05da84f1a03b7a10b70) Windows Server 2008 (not R2) and Windows Vista. All should work. :-)

**Comment by ar7z1** [comment]
RabbitMQ installation makes build time too slow: 10 minutes vs. 21 minutes.

slow

@dagwieers Maybe we should disable these tests from autorun?

**Comment by dagwieers** [comment]
But they are only run if these modules have been modified, I would expect. So it won't impact every run.

I let @mattclay decide on this one.

**Comment by ar7z1** [comment]
> But they are only run if these modules have been modified, I would expect.

Cool! :-)

**Comment by ar7z1** [comment]
@mattclay What do you think?

**Comment by mattclay** [comment]
The slow install shouldn't be a problem, since these tests will only run when changes specific to this module or tests are made. So yes, please include the tests for running in CI.

**Comment by ar7z1** [comment]
Cool! Then, I think, we can merge this PR. :-)

**Comment by jborean93** [comment]
Thanks @ar7z1, we can aways disable the tests from running in CI in the future if we need to.

**Pull request merged**